include/circt/Dialect/Calyx/CalyxStructure.td

```diff
@@ -9,3 +9,124 @@
 // This describes the Calyx structures.
 //
 //===----------------------------------------------------------------------===//
+
+def ComponentOp : CalyxOp<"component", [
+    Symbol,
+    FunctionLike,
+    IsolatedFromAbove,
+    // TODO(Calyx): Need to eventually expose the output ports.
+    NoTerminator,
+    SingleBlock
+  ]> {
+  let summary = "Calyx Component";
+  let description = [{
+    The "calyx.component" operation represents an overall
+    Calyx component containing:
+    (1) In- and output port definitions
+        that define the interface.
+    (2) The cells, wires, and control schedule.
+    ```mlir
+      component MyComponent(in1: 32) -> (out1: 32) {
+        cells { ... }
+        wires { ... }
+        control { ... }
+      }
+    ```
+  }];
+
+  // TODO(Calyx): Allow explicit port naming?
+  let arguments = (ins);
+  let results = (outs);
+
+  let regions = (region SizedRegion<1>: $body);
+
+  let extraClassDeclaration = [{
+    // Necessary to avoid name clashing with `front`.
+    using FunctionLike::front;
+
+    private:
+      // This trait needs access to the hooks defined below.
+      friend class OpTrait::FunctionLike<ComponentOp>;
+
+      /// Hooks for the input/output type enumeration in FunctionLike.
+      unsigned getNumFuncArguments() { return getType().getNumInputs(); }
+      unsigned getNumFuncResults() { return getType().getNumResults(); }
+    public:
+      /// Returns the body of a Calyx component.
+      Block *getBody() { return &getOperation()->getRegion(0).front(); }
+  }];
+
+  let printer = "return ::print$cppClass(p, *this);";
+  let parser = "return ::parse$cppClass(parser, result);";
+  let verifier = "return ::verify$cppClass(*this);";
+}
+
+def CellsOp : CalyxOp<"cells", [
+    IsolatedFromAbove,
+    SingleBlock,
+    NoTerminator,
+    HasParent<"ComponentOp">
+  ]> {
+  let summary = "Calyx Cells";
+  let description = [{
+    The "calyx.cells" operation represents a container for
+    the sub-components that are used within the parent
+    component.
+
+    ```mlir
+      cells {
+        adder = std_add(32);
+        myComp = MyComponent();
+      }
+    ```
+  }];
+  let regions = (region SizedRegion<1>: $body);
+  let printer = "return ::print$cppClass(p, *this);";
```

Comment:
I think CellsOp, WiresOp, and ControlOp can all use the declarative assembly format. I'm pretty sure it could just be:

```let assemblyFormat = "$body attr-dict";```

Then you could drop all the parsers and printers for those. In fact, you could probably define a base CalyxContainer class in ODS, since these are so similar, and I think are only varying in their traits and description.